src/fu-cabinet.c

```diff
@@ -877,6 +873,41 @@ fu_cabinet_parse(FuFirmware *firmware,
 	return TRUE;
 }
 
+GPtrArray *
+fu_cabinet_get_components(FuCabinet *self, GError **error)
+{
+	g_autoptr(GError) error_local = NULL;
+	g_autoptr(GPtrArray) components = NULL;
+
+	g_return_val_if_fail(FU_IS_CABINET(self), NULL);
+	g_return_val_if_fail(error == NULL || *error == NULL, NULL);
+
+	components =
+	    xb_silo_query(self->silo, "components/component[@type='firmware']", 0, &error_local);
+	if (components == NULL) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_INVALID_FILE,
+			    "no components: %s",
+			    error_local->message);
```

Comment:
do you really need the invalid file error code?  If not, you can discard all the error_local stuff

Reply:
Alas yes -- I think an audit of error domains is a good idea -- but not in this PR.